In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../data/processed/breast_cancer_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 569
Columns : 31


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 569
Columns : 30

TARGET VARIABLE (y)
Rows : 569


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


target
1    357
0    212
Name: count, dtype: int64


Class Percentage


target
1    62.74
0    37.26
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
mean radius,float64
mean texture,float64
mean perimeter,float64
mean area,float64
mean smoothness,float64
mean compactness,float64
mean concavity,float64
mean concave points,float64
mean symmetry,float64
mean fractal dimension,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (455, 30)
y_train shape : (455,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (114, 30)
y_test shape  : (114,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
1    62.64
0    37.36
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
1    63.16
0    36.84
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (455, 30)
X_test_scaled shape  : (114, 30)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,-1.072001,-0.658425,-1.088080,-0.939274,-0.135940,-1.008718,-0.968359,-1.102032,0.281062,-0.113231,...,-1.034094,-0.623497,-1.070773,-0.876534,-0.169982,-1.038836,-1.078995,-1.350527,-0.352658,-0.541380
1,1.748743,0.066502,1.751157,1.745559,1.274468,0.842288,1.519852,1.994664,-0.293045,-0.320180,...,1.228342,-0.092833,1.187467,1.104386,1.517001,0.249655,1.178594,1.549916,0.191078,-0.173739
2,-0.974734,-0.931124,-0.997709,-0.867589,-0.613515,-1.138154,-1.092292,-1.243358,0.434395,-0.429247,...,-0.973231,-1.036772,-1.008044,-0.834168,-1.097823,-1.167260,-1.282241,-1.707442,-0.307734,-1.213033
3,-0.145103,-1.215186,-0.123013,-0.253192,0.664482,0.286762,-0.129729,-0.098605,0.555635,0.029395,...,-0.251266,-1.369643,-0.166633,-0.330292,0.234006,0.096874,-0.087521,-0.344838,0.242198,-0.118266
4,-0.771617,-0.081211,-0.803700,-0.732927,-0.672282,-1.006099,-0.798502,-0.684484,0.737495,-0.457213,...,-0.801135,0.079230,-0.824381,-0.741830,-0.911367,-0.984612,-0.933190,-0.777604,0.555118,-0.761639


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


mean radius               -0.0
mean texture               0.0
mean perimeter            -0.0
mean area                 -0.0
mean smoothness            0.0
mean compactness          -0.0
mean concavity            -0.0
mean concave points        0.0
mean symmetry             -0.0
mean fractal dimension     0.0
radius error              -0.0
texture error             -0.0
perimeter error            0.0
area error                 0.0
smoothness error          -0.0
compactness error         -0.0
concavity error            0.0
concave points error      -0.0
symmetry error             0.0
fractal dimension error    0.0
worst radius               0.0
worst texture             -0.0
worst perimeter           -0.0
worst area                 0.0
worst smoothness           0.0
worst compactness         -0.0
worst concavity            0.0
worst concave points      -0.0
worst symmetry            -0.0
worst fractal dimension    0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


mean radius                1.0
mean texture               1.0
mean perimeter             1.0
mean area                  1.0
mean smoothness            1.0
mean compactness           1.0
mean concavity             1.0
mean concave points        1.0
mean symmetry              1.0
mean fractal dimension     1.0
radius error               1.0
texture error              1.0
perimeter error            1.0
area error                 1.0
smoothness error           1.0
compactness error          1.0
concavity error            1.0
concave points error       1.0
symmetry error             1.0
fractal dimension error    1.0
worst radius               1.0
worst texture              1.0
worst perimeter            1.0
worst area                 1.0
worst smoothness           1.0
worst compactness          1.0
worst concavity            1.0
worst concave points       1.0
worst symmetry             1.0
worst fractal dimension    1.0
dtype: float64

# MRMR

## 1. Baseline

In [10]:
!pip install mrmr-selection

In [11]:
# =============================================================================
# MRMR BASELINE FEATURE SELECTION
# TOP-K = 5, 10, 15, 20
# =============================================================================

from mrmr import mrmr_classif
import pandas as pd


# -------------------------------------------------------------------------
# Align indices
# -------------------------------------------------------------------------

X_train_mrmr = X_train_scaled.copy()
y_train_mrmr = y_train.copy()

X_train_mrmr = X_train_mrmr.reset_index(drop=True)
y_train_mrmr = y_train_mrmr.reset_index(drop=True)


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Store MRMR results
# -------------------------------------------------------------------------

mrmr_results = {}


# =============================================================================
# RUN MRMR
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"TOP {k} MRMR FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # MRMR feature selection
    # ---------------------------------------------------------------------

    selected_features = mrmr_classif(
        X=X_train_mrmr,
        y=y_train_mrmr,
        K=k
    )


    # ---------------------------------------------------------------------
    # Select features
    # ---------------------------------------------------------------------

    X_train_selected = X_train_scaled[
        selected_features
    ]

    X_test_selected = X_test_scaled[
        selected_features
    ]


    # ---------------------------------------------------------------------
    # Store results
    # ---------------------------------------------------------------------

    mrmr_results[k] = {

        "features": selected_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected

    }


    # ---------------------------------------------------------------------
    # Display
    # ---------------------------------------------------------------------

    print("\nSelected Features:")

    print(selected_features)


print("\n" + "=" * 70)
print("MRMR FEATURE SELECTION COMPLETED")
print("=" * 70)


TOP 5 MRMR FEATURES


100%|██████████| 5/5 [00:00<00:00, 31.37it/s]



Selected Features:
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter']

TOP 10 MRMR FEATURES


100%|██████████| 10/10 [00:00<00:00, 34.98it/s]



Selected Features:
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'worst concavity', 'mean concavity', 'mean area']

TOP 15 MRMR FEATURES


100%|██████████| 15/15 [00:00<00:00, 57.11it/s]



Selected Features:
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'worst concavity', 'mean concavity', 'mean area', 'worst compactness', 'mean compactness', 'worst texture', 'radius error', 'area error']

TOP 20 MRMR FEATURES


100%|██████████| 20/20 [00:00<00:00, 27.02it/s]


Selected Features:
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'worst concavity', 'mean concavity', 'mean area', 'worst compactness', 'mean compactness', 'worst texture', 'radius error', 'area error', 'worst smoothness', 'perimeter error', 'worst symmetry', 'mean texture', 'concave points error']

MRMR FEATURE SELECTION COMPLETED


In [12]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        scale_pos_weight=(
            (y_train == 0).sum() /
            (y_train == 1).sum()
        )
    )
}

In [13]:
# =============================================================================
# MRMR MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


mrmr_model_results = []


for k in k_values:

    Xtr = mrmr_results[k]["X_train"]

    Xts = mrmr_results[k]["X_test"]

    features = ", ".join(
        mrmr_results[k]["features"]
    )


    for model_name, model in models.items():

        # -------------------------------------------------------------
        # Train
        # -------------------------------------------------------------

        model.fit(
            Xtr,
            y_train
        )


        # -------------------------------------------------------------
        # Predictions
        # -------------------------------------------------------------

        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # -------------------------------------------------------------
        # Store metrics
        # -------------------------------------------------------------

        mrmr_model_results.append({

            "Method": "MRMR",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# =============================================================================
# CREATE RESULTS DATAFRAME
# =============================================================================

mrmr_results_df = pd.DataFrame(
    mrmr_model_results
)


display(mrmr_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,MRMR,5,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.929825,0.984848,0.902778,0.942029,0.991071
1,MRMR,5,Random Forest,"worst concave points, worst perimeter, mean co...",0.938596,0.971014,0.930556,0.950355,0.984788
2,MRMR,5,XGBoost,"worst concave points, worst perimeter, mean co...",0.938596,0.957746,0.944444,0.951049,0.984458
3,MRMR,10,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.912281,0.969697,0.888889,0.927536,0.991402
4,MRMR,10,Random Forest,"worst concave points, worst perimeter, mean co...",0.947368,0.958333,0.958333,0.958333,0.990575
5,MRMR,10,XGBoost,"worst concave points, worst perimeter, mean co...",0.964912,0.959459,0.986111,0.972603,0.991071
6,MRMR,15,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.938596,0.957746,0.944444,0.951049,0.994048
7,MRMR,15,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.958904,0.972222,0.965517,0.992890
8,MRMR,15,XGBoost,"worst concave points, worst perimeter, mean co...",0.956140,0.958904,0.972222,0.965517,0.996693
9,MRMR,20,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.964912,0.985714,0.958333,0.971831,0.995040


In [14]:
mrmr_results_df.to_csv(
    "../../../results/breast_cancer/baseline_results/mrmr_baseline_results.csv",
    index=False
)

## 2. DODA

In [15]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [16]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\johnm\msc_research\.venv\Scripts\python.exe
DODA: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\__init__.py
Selector: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\selector.py


In [17]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [18]:
# =============================================================================
# MRMR ADAPTER FOR DODA
# =============================================================================

import numpy as np


class MRMRAdapter:

    def __init__(self, K):
        self.K = K
        self.importance = {}

    def fit(self, X, y):

        # Run MRMR
        selected_features = mrmr_classif(
            X=X,
            y=y,
            K=self.K
        )

        # -------------------------------------------------------------
        # Convert MRMR ranking into importance scores
        # First feature = highest score
        # -------------------------------------------------------------

        n_features = len(selected_features)

        scores = np.zeros(len(X.columns))

        feature_to_index = {
            feature: i
            for i, feature in enumerate(X.columns)
        }

        for rank, feature in enumerate(selected_features):

            index = feature_to_index[feature]

            scores[index] = (
                n_features - rank
            ) / n_features

        # -------------------------------------------------------------
        # Store importance
        # -------------------------------------------------------------

        self.importance = dict(
            zip(
                X.columns,
                scores
            )
        )

        return self

    def get_importance(self):

        return self.importance

    def get_name(self):

        return "MRMR"

In [19]:
# ============================================================
# FIX INDEX ALIGNMENT FOR MRMR
# ============================================================

# Convert X and y to pandas objects
X_mrmr = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
).reset_index(drop=True)

y_mrmr = pd.Series(
    y_train
).reset_index(drop=True)


print("X shape:", X_mrmr.shape)
print("y shape:", y_mrmr.shape)

print("X index:", X_mrmr.index[:5])
print("y index:", y_mrmr.index[:5])

X shape: (455, 30)
y shape: (455,)
X index: RangeIndex(start=0, stop=5, step=1)
y index: RangeIndex(start=0, stop=5, step=1)


In [20]:
from mrmr import mrmr_classif

mrmr_ranked_features = mrmr_classif(
    X=X_mrmr,
    y=y_mrmr,
    K=X_mrmr.shape[1]
)

print("Full MRMR ranking:")
print(mrmr_ranked_features)

print("Number of features:", len(mrmr_ranked_features))

100%|██████████| 30/30 [00:08<00:00,  3.34it/s]

Full MRMR ranking:
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'worst concavity', 'mean concavity', 'mean area', 'worst compactness', 'mean compactness', 'worst texture', 'radius error', 'area error', 'worst smoothness', 'perimeter error', 'worst symmetry', 'mean texture', 'concave points error', 'mean smoothness', 'mean symmetry', 'worst fractal dimension', 'compactness error', 'concavity error', 'smoothness error', 'fractal dimension error', 'texture error', 'symmetry error', 'mean fractal dimension']
Number of features: 30


In [21]:
# =============================================================================
# MRMR + DODA FEATURE SELECTION
# =============================================================================

import numpy as np
import pandas as pd


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "../../../config/clinical_weights/breast_cancer.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# MRMR Adapter
# -------------------------------------------------------------------------

class MRMRAdapter:

    def __init__(self, ranked_features):

        self.ranked_features = ranked_features
        self.importance = {}


    def fit(self, X, y):

        n_features = len(self.ranked_features)

        # Convert MRMR ranking into decreasing scores
        # Rank 1 = highest score
        # Rank n = lowest score

        for rank, feature in enumerate(
            self.ranked_features,
            start=1
        ):

            self.importance[feature] = (
                n_features - rank + 1
            ) / n_features

        return self


    def get_importance(self):

        return self.importance


    def get_name(self):

        return "MRMR"


# -------------------------------------------------------------------------
# Store results
# -------------------------------------------------------------------------

mrmr_doda_results = {}


# =============================================================================
# RUN MRMR + DODA
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"MRMR + DODA \u2014 TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # Create MRMR adapter
    # ---------------------------------------------------------------------

    mrmr_adapter = MRMRAdapter(
        mrmr_ranked_features
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(

        operators=[
            mrmr_adapter
        ],

        provider=provider,

        fusion=fusion,

        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on training data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform test data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = (
        selector.get_selected_features()
    )


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store results
    # ---------------------------------------------------------------------

    mrmr_doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores":
            selector.raw_math_scores_,

        "math_scores":
            selector.math_scores_,

        "clinical_weights":
            selector.clinical_weights_,

        "final_scores":
            selector.final_scores_
    }


print("\n" + "=" * 70)
print("MRMR + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


MRMR + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x0000026B9A739400>
DODA Pipeline Started
Running: MRMR
Operator Scores:
{'MRMR_1': {'worst concave points': 1.0, 'worst perimeter': 0.9666666666666667, 'mean concave points': 0.9333333333333333, 'worst radius': 0.9, 'mean perimeter': 0.8666666666666667, 'worst area': 0.8333333333333334, 'mean radius': 0.8, 'worst concavity': 0.7666666666666667, 'mean concavity': 0.7333333333333333, 'mean area': 0.7, 'worst compactness': 0.6666666666666666, 'mean compactness': 0.6333333333333333, 'worst texture': 0.6, 'radius error': 0.5666666666666667, 'area error': 0.5333333333333333, 'worst smoothness': 0.5, 'perimeter error': 0.4666666666666667, 'worst symmetry': 0.43333333333333335, 'mean texture': 0.4, 'concave points error': 0.36666666666666664, 'mean smoothness': 0.3333333333333333, 'mean symmetry': 0.3, 'worst fractal dimension': 0.26666666666666666, 'compactness error

In [22]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = mrmr_doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
0,worst concave points,1.000000,1.000000
1,worst perimeter,0.966667,0.966667
2,mean concave points,0.933333,0.933333
3,worst radius,0.900000,0.900000
4,mean perimeter,0.866667,0.866667
5,worst area,0.833333,0.833333
6,mean radius,0.800000,0.800000
7,worst concavity,0.766667,0.766667
8,mean concavity,0.733333,0.733333
9,mean area,0.700000,0.700000


In [23]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
0,worst concave points,1.0
1,worst perimeter,1.0
3,worst radius,1.0
5,worst area,1.0
7,worst concavity,1.0
2,mean concave points,0.9
4,mean perimeter,0.9
6,mean radius,0.9
8,mean concavity,0.9
9,mean area,0.9


In [24]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
0,worst concave points,1.000000
1,worst perimeter,0.966667
3,worst radius,0.900000
2,mean concave points,0.840000
5,worst area,0.833333
4,mean perimeter,0.780000
7,worst concavity,0.766667
6,mean radius,0.720000
8,mean concavity,0.660000
9,mean area,0.630000


In [25]:
selector.get_selected_features()

['worst concave points',
 'worst perimeter',
 'worst radius',
 'mean concave points',
 'worst area',
 'mean perimeter',
 'worst concavity',
 'mean radius',
 'mean concavity',
 'mean area']

In [26]:
comparison = selector.compare_scores()

comparison

,Feature,Math Rank,Normalized Math Score,Clinical Weight,DODA Rank,Final Score,Rank Change
0,worst concave points,1,1.0000,1.0,1,1.0000,0
1,worst perimeter,2,0.9667,1.0,2,0.9667,0
2,worst radius,4,0.9000,1.0,3,0.9000,1
3,mean concave points,3,0.9333,0.9,4,0.8400,-1
4,worst area,6,0.8333,1.0,5,0.8333,1
5,mean perimeter,5,0.8667,0.9,6,0.7800,-1
6,worst concavity,8,0.7667,1.0,7,0.7667,1
7,mean radius,7,0.8000,0.9,8,0.7200,-1
8,mean concavity,9,0.7333,0.9,9,0.6600,0
9,mean area,10,0.7000,0.9,10,0.6300,0


In [27]:
# =============================================================================
# MRMR + DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


mrmr_doda_model_results = []


# =============================================================================
# EVALUATE EACH TOP-K
# =============================================================================

for k in k_values:

    Xtr = mrmr_doda_results[k]["X_train"]

    Xts = mrmr_doda_results[k]["X_test"]

    features = ", ".join(
        mrmr_doda_results[k]["features"]
    )


    # -------------------------------------------------------------------------
    # Evaluate each ML model
    # -------------------------------------------------------------------------

    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        # Predictions
        y_pred = model.predict(
            Xts
        )


        # Probability predictions
        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # ---------------------------------------------------------------------
        # Store results
        # ---------------------------------------------------------------------

        mrmr_doda_model_results.append({

            "Method": "MRMR + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# =============================================================================
# CREATE RESULTS DATAFRAME
# =============================================================================

mrmr_doda_results_df = pd.DataFrame(
    mrmr_doda_model_results
)


display(mrmr_doda_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,MRMR + DODA,5,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.938596,0.985075,0.916667,0.949640,0.990741
1,MRMR + DODA,5,Random Forest,"worst concave points, worst perimeter, worst r...",0.938596,0.971014,0.930556,0.950355,0.982804
2,MRMR + DODA,5,XGBoost,"worst concave points, worst perimeter, worst r...",0.921053,0.943662,0.930556,0.937063,0.982474
3,MRMR + DODA,10,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.912281,0.969697,0.888889,0.927536,0.991402
4,MRMR + DODA,10,Random Forest,"worst concave points, worst perimeter, worst r...",0.947368,0.958333,0.958333,0.958333,0.988591
5,MRMR + DODA,10,XGBoost,"worst concave points, worst perimeter, worst r...",0.964912,0.959459,0.986111,0.972603,0.991071
6,MRMR + DODA,15,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.938596,0.957746,0.944444,0.951049,0.994048
7,MRMR + DODA,15,Random Forest,"worst concave points, worst perimeter, worst r...",0.938596,0.957746,0.944444,0.951049,0.993056
8,MRMR + DODA,15,XGBoost,"worst concave points, worst perimeter, worst r...",0.947368,0.958333,0.958333,0.958333,0.995370
9,MRMR + DODA,20,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.956140,0.985507,0.944444,0.964539,0.994709


In [28]:
mrmr_doda_results_df.to_csv(
    "../../../results/breast_cancer/doda_results/mrmr_doda_results.csv",
    index=False
)

In [29]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        mrmr_results_df,
        mrmr_doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,MRMR,5,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.929825,0.984848,0.902778,0.942029,0.991071
1,MRMR,5,Random Forest,"worst concave points, worst perimeter, mean co...",0.938596,0.971014,0.930556,0.950355,0.984788
2,MRMR,5,XGBoost,"worst concave points, worst perimeter, mean co...",0.938596,0.957746,0.944444,0.951049,0.984458
3,MRMR,10,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.912281,0.969697,0.888889,0.927536,0.991402
4,MRMR,10,Random Forest,"worst concave points, worst perimeter, mean co...",0.947368,0.958333,0.958333,0.958333,0.990575
5,MRMR,10,XGBoost,"worst concave points, worst perimeter, mean co...",0.964912,0.959459,0.986111,0.972603,0.991071
6,MRMR,15,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.938596,0.957746,0.944444,0.951049,0.994048
7,MRMR,15,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.958904,0.972222,0.965517,0.992890
8,MRMR,15,XGBoost,"worst concave points, worst perimeter, mean co...",0.956140,0.958904,0.972222,0.965517,0.996693
9,MRMR,20,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.964912,0.985714,0.958333,0.971831,0.995040


In [30]:
for k in k_values:

    selector = mrmr_doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'worst concavity', 'mean concavity', 'mean area', 'worst compactness', 'mean compactness', 'worst texture', 'radius error', 'area error', 'worst smoothness', 'perimeter error', 'worst symmetry', 'mean texture', 'concave points error', 'mean smoothness', 'mean symmetry', 'worst fractal dimension', 'compactness error', 'concavity error', 'smoothness error', 'fractal dimension error', 'texture error', 'symmetry error', 'mean fractal dimension']

DODA Ranking:
['worst concave points', 'worst perimeter', 'worst radius', 'mean concave points', 'worst area', 'mean perimeter', 'worst concavity', 'mean radius', 'mean concavity', 'mean area', 'worst compactness', 'mean compactness', 'worst texture', 'radius error', 'area error', 'worst symmetry', 'worst smoothness', 'perimeter error', 'mean texture', 'mean symmetry', 'concave points error', 'm

# 3. mRMR + DODA: Model Evaluation & Interpretation

## 3.1 Overall Performance

Both mRMR and mRMR + DODA produced strong predictive performance across the tested Top-K feature subsets. Most configurations achieved accuracy and F1-scores above 0.92, with ROC-AUC values generally close to or above 0.98.

The best mRMR baseline configuration was **Logistic Regression with Top-K = 20**, which achieved the highest baseline accuracy and F1-score, while XGBoost reached the highest baseline ROC-AUC at Top-K = 15.

The best mRMR + DODA configurations were **XGBoost with Top-K = 10 and Top-K = 20** for accuracy and F1-score, with XGBoost at Top-K = 20 also matching the highest DODA ROC-AUC.

DODA did **not** consistently improve predictive performance. Its effect was configuration-dependent: it improved Logistic Regression at Top-K = 5, maintained several Top-K = 10 and Top-K = 15 results, and reduced performance in some Random Forest and XGBoost settings. The main trade-off was that clinical weighting refined the feature ranking, but this did not always translate into higher model performance.


## 3.2 mRMR Baseline

### Top-K = 5

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC |
|---|---:|---:|---:|---:|---:|
| Logistic Regression | 0.929825 | 0.984848 | 0.902778 | 0.942029 | 0.991071 |
| Random Forest | 0.938596 | 0.971014 | 0.930556 | 0.950355 | 0.984788 |
| XGBoost | 0.938596 | 0.957746 | 0.944444 | 0.951049 | 0.984458 |

At Top-K = 5, XGBoost gave the strongest F1-score, while Logistic Regression achieved the highest ROC-AUC.

### Top-K = 10

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC |
|---|---:|---:|---:|---:|---:|
| Logistic Regression | 0.912281 | 0.969697 | 0.888889 | 0.927536 | 0.991402 |
| Random Forest | 0.947368 | 0.958333 | 0.958333 | 0.958333 | 0.990575 |
| XGBoost | 0.964912 | 0.959459 | 0.986111 | 0.972603 | 0.991071 |

At Top-K = 10, XGBoost was the strongest model by accuracy, recall, and F1-score. Logistic Regression had a marginally higher ROC-AUC.

### Top-K = 15

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC |
|---|---:|---:|---:|---:|---:|
| Logistic Regression | 0.938596 | 0.957746 | 0.944444 | 0.951049 | 0.994048 |
| Random Forest | 0.956140 | 0.958904 | 0.972222 | 0.965517 | 0.992890 |
| XGBoost | 0.956140 | 0.958904 | 0.972222 | 0.965517 | 0.996693 |

At Top-K = 15, Random Forest and XGBoost had equal accuracy, recall, and F1-score, while XGBoost achieved the highest ROC-AUC.

### Top-K = 20

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC |
|---|---:|---:|---:|---:|---:|
| Logistic Regression | 0.964912 | 0.985714 | 0.958333 | 0.971831 | 0.995040 |
| Random Forest | 0.956140 | 0.971831 | 0.958333 | 0.965035 | 0.994378 |
| XGBoost | 0.956140 | 0.958904 | 0.972222 | 0.965517 | 0.996032 |

At Top-K = 20, Logistic Regression was the strongest baseline model by accuracy, precision, and F1-score. XGBoost achieved the highest ROC-AUC.


## 3.3 mRMR + DODA

The table below compares the main performance metrics between the mRMR baseline and mRMR + DODA.

| Top-K | Model | Metric | mRMR | mRMR + DODA | Effect |
|------:|-------|--------|-----:|------------:|:------:|
| 5 | Logistic Regression | F1-score | 0.942029 | 0.949640 | Improved |
| 5 | Logistic Regression | ROC-AUC | 0.991071 | 0.990741 | Maintained / Comparable |
| 5 | Random Forest | F1-score | 0.950355 | 0.950355 | Maintained / Comparable |
| 5 | Random Forest | ROC-AUC | 0.984788 | 0.982804 | Reduced |
| 5 | XGBoost | F1-score | 0.951049 | 0.937063 | Reduced |
| 5 | XGBoost | ROC-AUC | 0.984458 | 0.982474 | Reduced |
| 10 | Logistic Regression | F1-score | 0.927536 | 0.927536 | Maintained / Comparable |
| 10 | Logistic Regression | ROC-AUC | 0.991402 | 0.991402 | Maintained / Comparable |
| 10 | Random Forest | F1-score | 0.958333 | 0.958333 | Maintained / Comparable |
| 10 | Random Forest | ROC-AUC | 0.990575 | 0.988591 | Reduced |
| 10 | XGBoost | F1-score | 0.972603 | 0.972603 | Maintained / Comparable |
| 10 | XGBoost | ROC-AUC | 0.991071 | 0.991071 | Maintained / Comparable |
| 15 | Logistic Regression | F1-score | 0.951049 | 0.951049 | Maintained / Comparable |
| 15 | Logistic Regression | ROC-AUC | 0.994048 | 0.994048 | Maintained / Comparable |
| 15 | Random Forest | F1-score | 0.965517 | 0.951049 | Reduced |
| 15 | Random Forest | ROC-AUC | 0.992890 | 0.993056 | Maintained / Comparable |
| 15 | XGBoost | F1-score | 0.965517 | 0.958333 | Reduced |
| 15 | XGBoost | ROC-AUC | 0.996693 | 0.995370 | Reduced |
| 20 | Logistic Regression | F1-score | 0.971831 | 0.964539 | Reduced |
| 20 | Logistic Regression | ROC-AUC | 0.995040 | 0.994709 | Maintained / Comparable |
| 20 | Random Forest | F1-score | 0.965035 | 0.957746 | Reduced |
| 20 | Random Forest | ROC-AUC | 0.994378 | 0.994048 | Maintained / Comparable |
| 20 | XGBoost | F1-score | 0.965517 | 0.972222 | Improved |
| 20 | XGBoost | ROC-AUC | 0.996032 | 0.996032 | Maintained / Comparable |

Overall, DODA produced a mixed pattern. It improved selected configurations, especially Logistic Regression at Top-K = 5 and XGBoost F1-score at Top-K = 20, but it also reduced performance for several Random Forest and XGBoost configurations. Therefore, the executed results do not support a conclusion that DODA consistently improves mRMR predictive performance.


## 3.4 Effect on Top-K Feature Selection

The mRMR baseline ranking began with:

| Rank | mRMR Feature |
|---:|---|
| 1 | worst concave points |
| 2 | worst perimeter |
| 3 | mean concave points |
| 4 | worst radius |
| 5 | mean perimeter |
| 6 | worst area |
| 7 | mean radius |
| 8 | worst concavity |
| 9 | mean concavity |
| 10 | mean area |

The mRMR + DODA ranking began with:

| Rank | mRMR + DODA Feature |
|---:|---|
| 1 | worst concave points |
| 2 | worst perimeter |
| 3 | worst radius |
| 4 | mean concave points |
| 5 | worst area |
| 6 | mean perimeter |
| 7 | worst concavity |
| 8 | mean radius |
| 9 | mean concavity |
| 10 | mean area |

Important high-ranked features were retained, including `worst concave points`, `worst perimeter`, `mean concave points`, `worst radius`, `mean perimeter`, `worst area`, `mean radius`, `worst concavity`, `mean concavity`, and `mean area`.

DODA mainly refined the ordering rather than substantially changing the Top-K = 10 subset. In the Top-K = 5 subset, `worst radius` and `worst area` moved upward, while `mean perimeter` left the Top-5. At Top-K = 20, `mean symmetry` entered the DODA-selected subset, while `concave points error` was no longer included.

The upward movement of features such as `worst radius`, `worst area`, and `worst concavity` is aligned with their higher configured clinical weights in the executed outputs. However, this should be interpreted as ranking refinement rather than direct evidence of clinical improvement. The final DODA scores also show some repeated zero values among lower-ranked features, so the ordering of those low-scoring features should be interpreted cautiously.


## 3.5 Model-Level Comparison

### Logistic Regression

Logistic Regression improved at Top-K = 5, was maintained at Top-K = 10 and Top-K = 15, and was reduced at Top-K = 20 for accuracy, recall, and F1-score. Its best DODA configuration was Top-K = 20, although the mRMR baseline at Top-K = 20 remained slightly stronger. Overall, DODA had a mostly comparable effect for Logistic Regression, with one improvement at the smallest feature subset.

### Random Forest

Random Forest performance was maintained at Top-K = 5 and Top-K = 10 for accuracy and F1-score, but reduced at Top-K = 15 and Top-K = 20. Its best DODA configuration was Top-K = 10. Overall, DODA did not improve Random Forest performance in the executed results.

### XGBoost

XGBoost was reduced at Top-K = 5 and Top-K = 15, maintained at Top-K = 10, and improved in F1-score at Top-K = 20. Its best DODA configurations were Top-K = 10 and Top-K = 20, both reaching 0.964912 accuracy, with Top-K = 20 achieving the highest DODA ROC-AUC. Overall, DODA had a configuration-dependent effect for XGBoost rather than a consistently positive one.


## 3.6 Key Findings

| Finding | Observation |
|---------|-------------|
| Best mRMR configuration | Logistic Regression with Top-K = 20 achieved 0.964912 accuracy and 0.971831 F1-score. XGBoost at Top-K = 15 achieved the highest baseline ROC-AUC of 0.996693. |
| Best mRMR + DODA configuration | XGBoost with Top-K = 10 and Top-K = 20 achieved 0.964912 accuracy. XGBoost with Top-K = 20 achieved the highest DODA F1-score of 0.972222 and ROC-AUC of 0.996032. |
| Strongest DODA improvement | Logistic Regression at Top-K = 5 improved in accuracy, recall, and F1-score. XGBoost at Top-K = 20 also improved in accuracy, precision, and F1-score. |
| Largest performance reduction | XGBoost at Top-K = 5 showed the largest F1-score reduction, from 0.951049 to 0.937063. |
| Most stable model | Logistic Regression was the most stable across DODA comparisons, with several maintained results and only a small reduction at Top-K = 20. |
| Main feature-ranking effect | DODA moved clinically weighted `worst` features such as `worst radius`, `worst area`, and `worst concavity` upward in the ranking. |
| Overall DODA effect | DODA refined the mRMR ranking but did not consistently improve predictive performance across all models and Top-K values. |


## 3.7 Important Observations

- Predictive performance was strong for both approaches, with most configurations producing high F1-scores and ROC-AUC values.
- DODA's effect depended on both the model and the Top-K feature subset.
- The Top-K = 10 DODA results were especially stable, with Logistic Regression and XGBoost matching the mRMR baseline and Random Forest maintaining F1-score.
- DODA changed the ranking more clearly than it changed the overall selected feature set, especially for the Top-K = 10 and Top-K = 15 subsets.
- Clinical weighting increased the priority of several high-weight `worst` diagnostic features, but the resulting ranking changes did not guarantee better predictive performance.
- The experiment supports cautious interpretation only; it does not establish that clinical knowledge universally improves mRMR-based classification.


## 3.8 Conclusion

DODA did not consistently improve mRMR predictive performance in this breast cancer experiment. Some configurations improved, particularly Logistic Regression at Top-K = 5 and XGBoost at Top-K = 20, but other configurations were maintained or reduced.

DODA preserved competitive performance in several cases, especially at Top-K = 10, where multiple metrics remained unchanged or closely comparable to the mRMR baseline. This suggests that clinical weighting can be introduced without necessarily degrading performance in some settings.

DODA meaningfully altered the feature ranking by promoting clinically weighted features such as `worst radius`, `worst area`, and `worst concavity`. These changes were aligned with the configured clinical weights shown in the executed outputs. However, the results do not support a claim that these ranking changes directly caused clinical or predictive improvement.

Overall, the evidence suggests that clinical knowledge acted mainly as a refinement of statistically selected mRMR features rather than as a universally performance-improving mechanism. The experiment supports the value of DODA as an interpretable ranking adjustment method, but its predictive benefit remains model- and Top-K-dependent.
